[Style] Convert grid-{column,row}-{start,end} to use strong style types

weinig opened the pull request:
#### dbb0c1ea51ecf0059024ddde71cb0e3d300d7b7d
<pre>
[Style] Convert grid-{column,row}-{start,end} to use strong style types
<a href="https://bugs.webkit.org/show_bug.cgi?id=295906">https://bugs.webkit.org/show_bug.cgi?id=295906</a>

Reviewed by Darin Adler.

Converts `GridPosition` to a strong style type, enforcing type validity at construction
via sub types `GridPosition::Explicit` and `GridPosition::Span`.

Also:
  - Moves and updates `GridPositionsResolver` to utilize type switching
  - Moves `GridTrackSizingDirection` enum to its own header and makes it
    it a standard two item enum, not an option set style enum. The one
    use of its option set capability was switched to its own type.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Grid.cpp:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/inspector/InspectorOverlay.cpp:
* Source/WebCore/rendering/AncestorSubgridIterator.cpp:
* Source/WebCore/rendering/AncestorSubgridIterator.h:
* Source/WebCore/rendering/Grid.cpp:
* Source/WebCore/rendering/Grid.h:
* Source/WebCore/rendering/GridBaselineAlignment.cpp:
* Source/WebCore/rendering/GridBaselineAlignment.h:
* Source/WebCore/rendering/GridLayoutFunctions.cpp:
* Source/WebCore/rendering/GridLayoutFunctions.h:
* Source/WebCore/rendering/GridMasonryLayout.cpp:
* Source/WebCore/rendering/GridMasonryLayout.h:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderGrid.cpp:
* Source/WebCore/rendering/RenderGrid.h:
* Source/WebCore/rendering/RenderTheme.cpp:
* Source/WebCore/rendering/style/GridSpan.cpp:
* Source/WebCore/rendering/style/NameScope.h:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleGridData.cpp:
* Source/WebCore/rendering/style/StyleGridItemData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationFunctions.h:
* Source/WebCore/style/values/grid/StyleGridPosition.cpp: Renamed from Source/WebCore/rendering/style/GridPositions.h.
* Source/WebCore/style/values/grid/StyleGridPosition.h: Renamed from Source/WebCore/rendering/style/GridPositions.h.
* Source/WebCore/style/values/grid/StyleGridPositionSide.h: Added.
* Source/WebCore/style/values/grid/StyleGridPositionsResolver.cpp: Renamed from Source/WebCore/rendering/style/GridPositionsResolver.cpp.
* Source/WebCore/style/values/grid/StyleGridPositionsResolver.h: Renamed from Source/WebCore/rendering/style/GridPositionsResolver.h.
* Source/WebCore/style/values/grid/StyleGridTemplateAreas.cpp:
* Source/WebCore/style/values/grid/StyleGridTrackSizingDirection.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Logging.h:
* Source/WebCore/testing/Internals.cpp:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/GridPosition.cpp:

Canonical link: <a href="https://commits.webkit.org/297471@main">https://commits.webkit.org/297471@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b176685250d3bbfe6b9d64af312dc3a9c3093c50

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111821 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31483 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21960 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117839 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62053 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113784 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32165 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40066 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84945 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35638 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114768 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25682 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100628 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65441 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25008 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18764 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61684 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95063 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18836 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121194 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38851 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28882 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93872 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39231 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96887 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93696 "Found 2 new API test failures: /WebKitGTK/TestWebProcessExtensions:/webkit/WebKitWebView/web-process-crashed, /WebKitGTK/TestWebKitPolicyClient:/webkit/WebKitPolicyClient/new-window-policy (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23912 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38819 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16603 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34871 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38748 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44243 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38390 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41715 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40103 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/Grid.cpp:
```diff
@@ -36,19 +37,19 @@ Grid::Grid(RenderGrid& grid)
-    ASSERT(static_cast<int>(maximumColumnSize) < GridPosition::max() * 2);
-    const size_t oldColumnSize = numTracks(GridTrackSizingDirection::ForColumns);
-    const size_t oldRowSize = numTracks(GridTrackSizingDirection::ForRows);
+    ASSERT(static_cast<int>(maximumRowSize) < Style::GridPosition::max() * 2);
+    ASSERT(static_cast<int>(maximumColumnSize) < Style::GridPosition::max() * 2);
```
Not new to this patch: Seems like casting to `int` here is not good. Instead we should cast the other side to `unsigned`?

weinig replied:
Going to leave this one alone. I tried changing some things to unsigned in the grid code and things quickly got out of hand.

darinadler commented on Source/WebCore/rendering/Grid.cpp:
```diff
@@ -36,19 +37,19 @@ Grid::Grid(RenderGrid& grid)
+    ASSERT(static_cast<int>(maximumRowSize) < Style::GridPosition::max() * 2);
+    ASSERT(static_cast<int>(maximumColumnSize) < Style::GridPosition::max() * 2);
+    const size_t oldColumnSize = numTracks(Style::GridTrackSizingDirection::Columns);
+    const size_t oldRowSize = numTracks(Style::GridTrackSizingDirection::Rows);
```
Not new to this patch: We typically don’t use `const` for local variable scalars in WebKit.
```suggestion
    size_t oldColumnSize = numTracks(Style::GridTrackSizingDirection::Columns);
    size_t oldRowSize = numTracks(Style::GridTrackSizingDirection::Rows);
```

csaavedra commented on Source/WebCore/rendering/Grid.h:
```diff
@@ -57,15 +59,15 @@ class Grid final {
 
-    GridSpan gridItemSpan(const RenderBox&, GridTrackSizingDirection) const;
-    GridSpan gridItemSpanIgnoringCollapsedTracks(const RenderBox&, GridTrackSizingDirection) const;
+    GridSpan gridItemSpan(const RenderBox&, Style::GridTrackSizingDirection) const;
+    GridSpan gridItemSpanIgnoringCollapsedTracks(const RenderBox&, Style::GridTrackSizingDirection) const;
```
You're gonna need to forward-declare GridSpan in this header file to prevent non-unified builds from breaking.

csaavedra replied:
Told you so :) https://build.webkit.org/#/builders/133/builds/50387/steps/10/logs/errors

csaavedra reviewed (commented):
Not a reviewer, just a minor comment.

weinig committed:
[Style] Convert grid-{column,row}-{start,end} to use strong style types
https://bugs.webkit.org/show_bug.cgi?id=295906

Reviewed by Darin Adler.

Converts `GridPosition` to a strong style type, enforcing type validity at construction
via sub types `GridPosition::Explicit` and `GridPosition::Span`.

Also:
  - Moves and updates `GridPositionsResolver` to utilize type switching
  - Moves `GridTrackSizingDirection` enum to its own header and makes it
    it a standard two item enum, not an option set style enum. The one
    use of its option set capability was switched to its own type.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Grid.cpp:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/inspector/InspectorOverlay.cpp:
* Source/WebCore/rendering/AncestorSubgridIterator.cpp:
* Source/WebCore/rendering/AncestorSubgridIterator.h:
* Source/WebCore/rendering/Grid.cpp:
* Source/WebCore/rendering/Grid.h:
* Source/WebCore/rendering/GridBaselineAlignment.cpp:
* Source/WebCore/rendering/GridBaselineAlignment.h:
* Source/WebCore/rendering/GridLayoutFunctions.cpp:
* Source/WebCore/rendering/GridLayoutFunctions.h:
* Source/WebCore/rendering/GridMasonryLayout.cpp:
* Source/WebCore/rendering/GridMasonryLayout.h:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderGrid.cpp:
* Source/WebCore/rendering/RenderGrid.h:
* Source/WebCore/rendering/RenderTheme.cpp:
* Source/WebCore/rendering/style/GridSpan.cpp:
* Source/WebCore/rendering/style/NameScope.h:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleGridData.cpp:
* Source/WebCore/rendering/style/StyleGridItemData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationFunctions.h:
* Source/WebCore/style/values/grid/StyleGridPosition.cpp: Renamed from Source/WebCore/rendering/style/GridPositions.h.
* Source/WebCore/style/values/grid/StyleGridPosition.h: Renamed from Source/WebCore/rendering/style/GridPositions.h.
* Source/WebCore/style/values/grid/StyleGridPositionSide.h: Added.
* Source/WebCore/style/values/grid/StyleGridPositionsResolver.cpp: Renamed from Source/WebCore/rendering/style/GridPositionsResolver.cpp.
* Source/WebCore/style/values/grid/StyleGridPositionsResolver.h: Renamed from Source/WebCore/rendering/style/GridPositionsResolver.h.
* Source/WebCore/style/values/grid/StyleGridTemplateAreas.cpp:
* Source/WebCore/style/values/grid/StyleGridTrackSizingDirection.h: Added.
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Logging.h:
* Source/WebCore/testing/Internals.cpp:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/GridPosition.cpp:

Canonical link: https://commits.webkit.org/297471@main
